### ISSUE TITLE
Mouse-related Crash on Allegro programs

### ISSUE THREAD

**Issue by spagbolninja** [dosbox-staging/dosbox-staging#3987]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Debian

### Relevant hardware info

Thinkpad X201

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

(I ticked the above but was asked to file a new ticket as I had originally responded to https://github.com/dosbox-staging/dosbox-staging/issues/2180 )

Allegro's 4.2.3's 'test.exe' program crashes to DosBox Staging's command line (ver. 0.81.2) with a SIGSEGV error whenever the mouse touches the screen edge or Staging's window.

This is the pre-compiled version of DJGPP and Allegro made available at https://www.mrdictionary.net/allegro/ which provides a fix for a long standing soundblaster card issue. The above ticket (2180) uses a slightly different version still available from Allegro's site.

This crash also occurs for Allegro's 'setup.exe' which is used to detect a soundcard and test it with midi and wav sounds. Should the mouse his the screen edge or window border, it will crash to command line.

This is a very similar to the behavior expressed in the above linked ticket. There are other threads on forums expressing similar issues when dealing with programs/games that use the Allegro libraries in a dosbox fork. People have reported that the problem is not reproduced on original hardware or FreeDOS and is a phenomenon found in dosbox and its forks.

Experimentation with using a different mouse driver (CTmouse) screen aspect ratios, memory sizes etc. in the config file have not produced any changes to the behavior.

Following the suggestions in ticket 2180, I have tested it with DosBox-X and found that while the issue does occur at first, subsequently opening the programs from within the same DosBox-X session command line eliminates the program. **This is different to Staging where no matter how many times it is opened within the same Staging session it will crash back to the command line when the mouse touches a border.** This suggests to me that there is something different in the way that DosBox-X handles its mouse that at least partially solves the issue, and that likewise there is a difference here between the emulators and original hardware.

Thank you, hope this all helps!

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Download the test.exe file
2. Open test.exe in Dosbox staging
3. Move mouse cursor to any border of window (in windowed mode) or edge of rendered screen (fullscreen mode). Crash should occur immediately.

### Download URL of affected game or software

https://github.com/spagbolninja/testrepository/raw/refs/heads/main/TEST.EXE

### Your configuration

```ini
https://github.com/spagbolninja/testrepository/blob/main/dosbox-staging%20(spagbol%20moues%20crash).conf
```

### Provide a Log

```text
Console log:
https://github.com/spagbolninja/testrepository/blob/main/newlog.txt

Crash message in staging:
https://github.com/spagbolninja/testrepository/blob/main/2024-10-14_17-24.png?raw=true
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
Thanks @spagbolninja 🙏🏻

**Comment by spagbolninja** [comment]
Thought I better put this here too in case Allegro's DOS mouse module (DMOUSE.C) provides any clues as to what's happening: (https://github.com/spagbolninja/testrepository/blob/main/DMOUSE.C)

**Comment by FeralChild64** [comment]
> This suggests to me that there is something different in the way that DosBox-X handles its mouse that at least partially solves the issue, and that likewise there is a difference here between the emulators and original hardware.

It seems very unlikely to me that bug is within the mouse emulation code. If you run CtMouse (or any other DOS mouse driver), the emulation in DOSBox Staging will go through a completely different code path and the emulator won't even know whether the mouse cursor reached the screen border.

According to CWSDPMI documentation, https://github.com/jayschwa/cwsdpmi/blob/master/bin/cwsdpmi.doc:

> "General Protection Fault" - bad parameter sent to a DPMI call

This is the line that prints the error code: https://github.com/jayschwa/cwsdpmi/blob/ed2b39b48f12de16fc01aa2415a6b3c5b9531abc/src/control.c#L249

```c
  errmsg("ebp=%08lx esp=%08lx cs=%x ds=%x es=%x fs=%x gs=%x ss=%x error=%04x\n",
	tss_ptr->tss_ebp, tss_ptr->tss_esp,
	tss_ptr->tss_cs, tss_ptr->tss_ds, tss_ptr->tss_es, tss_ptr->tss_fs,
	tss_ptr->tss_gs, tss_ptr->tss_ss, (word16)tss_ptr->tss_error);
```

The `tss_error` is set either here, https://github.com/jayschwa/cwsdpmi/blob/ed2b39b48f12de16fc01aa2415a6b3c5b9531abc/src/tables.asm#L95:

```asm
	je	check_SW_int		; We don't have selectors this large !
	mov	eax,fs:[esi]
	mov	[bx].tss_error,eax
```

or here, https://github.com/jayschwa/cwsdpmi/blob/ed2b39b48f12de16fc01aa2415a6b3c5b9531abc/src/tables.asm#L141:

```asm
	mov	bx,_tss_ptr
	mov	[bx].tss_irqn,8		; double fault
	pop	[bx].tss_error		; dword
```

Of course I might be wrong, but to me it really looks more like a CPU/DPMI emulation problem than mouse emulation problem.

**Comment by Python-Exoproject** [comment]
@johnnovak Ive got another game to add to the list of Allegro games that cause a GPF when the mouse goes to the edge of the screen. The game is Biliardo 2 (included in eXoDOS) and below is the error. As well as when the mouse touches the edge of the screen you can also trigger a GPF at random times just playing the game but touching the edge guarantees it. If you prefer a new ticket let me know and ill make one

![Image](https://github.com/user-attachments/assets/533631d3-1b98-4a7a-9dd7-2acd167030e6)

**Comment by Python-Exoproject** [comment]
ripsaw8080 has also commented on this issue in relation to another program using Allegro, may or may not be relevant

https://www.vogons.org/viewtopic.php?p=983595#p983595
https://www.vogons.org/viewtopic.php?p=983696#p983696

**Comment by johnnovak** [comment]
@Python-Exoproject Thanks, this ticket is sufficient as the issue seems to be common to all Allegro games.

**Comment by FeralChild64** [comment]
And the problem is not specific to Staging.

**Comment by Python-Exoproject** [comment]
No the issue isnt specific to Staging, but it would be great if the provided info is enough for Staging to fix a longstanding emulation issue

If it helps isolate the problem, Zelda Classic (2000) is another Allegro game with the issue

![Image](https://github.com/user-attachments/assets/99b638d4-a50d-4a90-93a2-44942af91148)

**Comment by Python-Exoproject** [comment]
@johnnovak Huh, just tested using dosbox X and whatever the issue is they may have solved it as I didnt get a crash from either Biliardo 2, Zelda Classic, or the test.exe provided with the Allegro source. So their code holds the answer

**Comment by Python-Exoproject** [comment]
Isolated further. In X a change first made in Dosbox X 82.17 fixes the GPF crash, as their 82.16 crashes like other forks

**Comment by FeralChild64** [comment]
https://github.com/joncampbell123/dosbox-x/compare/dosbox-x-v0.82.16...dosbox-x-v0.82.17

If you still have the environment for testing, could you check if this is the `use dynamic core with paging on` option which fixes the problem? They've changed the default value between these two releases.

**Comment by Python-Exoproject** [comment]
Yep thats it, `use dynamic core with paging on=false` causes a crash as per this ticket while `use dynamic core with paging on=true` allows the games to run

**Comment by FeralChild64** [comment]
OK, so it looks like another CPU paging related problem, similar to https://github.com/dosbox-staging/dosbox-staging/issues/4097 and https://github.com/dosbox-staging/dosbox-staging/issues/4163.

Thanks for bisecting this!

**Comment by Python-Exoproject** [comment]
ooohhh interesting.

As you suggested above I have confirmed that the flag is the difference between crashing and working for those two tickets as well.

So the biggest issue im hitting with Staging and using it for Win3x is related to the same issue.

Is this something Staging can implement a fix for? Maybe a flag like X has or some other solution? I have a sneaking suspicion that as we try to convert games to Staging for eXoWin3x we are going to hit this issue a lot.

**Comment by FeralChild64** [comment]
It's not easy - the changes to port are rather huge and complex. One guy has already tried to port them, so far without success: https://github.com/dosbox-staging/dosbox-staging/compare/main...il98/page-fault-import-from-dosbox-pure-and-x
